### ISSUE TITLE
[readsf~] bug: reading some files never stops

### ISSUE THREAD

**Issue by Ant1r** [pure-data/pure-data#1786]
This is (I think) a long-standing bug in [readsf~], I'm quite sure I've encountered it before:
- open a file in [readsf~]
- start playback with [1(
- wait for a bang from [readsf~]'s right outlet... after a while, decide to do something else!

For this report I wanted to find such a wav file reproducing the bug, and I wanted the file to be short enough. The file has been built with Audacity and "damaged" by an Android java code (on which I'm currently working). In the attached test, `batt.wav` is the initial file, and plays normally; `batt_bad.wav` is the damaged one, and loops indefinitely.
[bug_readsf.zip](https://github.com/pure-data/pure-data/files/9819180/bug_readsf.zip)

I think two things are needed for the bug to happen:
- the file needs a special duration, e.g the bug happens with  both 529151 and 4863 stereo frames, i.e (0x81300 - 1) and (0x1300 - 1)
- the file's header has to be damaged in some way; here for instance, for the `batt_bad.wav` file:
```
 datasize calculated from the file size:              19456
 datasize calculated from the head 'riff_size' field: 19455
 datasize calculated from the head 'data_size' field: 19455
```
![readsf_bug](https://user-images.githubusercontent.com/3748246/196660960-994a2822-c627-4c03-9462-ad8af30e70ee.gif)

**Comment by umlaeute** [comment]
```sh
$ sndfile-info batt_bad.wav 
========================================
File : batt_bad.wav
Length : 19500
RIFF : 19491
WAVE
fmt  : 16
  Format        : 0x1 => WAVE_FORMAT_PCM
  Channels      : 2
  Sample Rate   : 44100
  Block Align   : 4
  Bit Width     : 16
  Bytes/sec     : 176400
*** 'data' chunk should be an even number of bytes in length.
data : 19455
End

----------------------------------------
Sample Rate : 44100
Frames      : 4863
Channels    : 2
Format      : 0x00010002
Sections    : 1
Seekable    : TRUE
Duration    : 00:00:00.110
Signal Max  : 24556 (-2.51 dB)
```

the important part is:
> *** 'data' chunk should be an even number of bytes in length.

**Comment by Ant1r** [comment]
> 'data' chunk should be an even number of bytes in length.

yes, I found that. Thanks anyway for mentioning `sndfile-info`, I had found `sfinfo` but it's not as performant.
(I still have to find out why my Java thing is doing that, but that's another story)

But I do think `[readsf~]` should never loop, in any case!

**Comment by umlaeute** [comment]
yes.
acctually, i think the error-message by `sndfile-info` is a bug in libsndfile (https://github.com/libsndfile/libsndfile/issues/887), so we are in good company.

**Comment by Ant1r** [comment]
> i think the error-message emitted by `sndfile-info` is a bug

? I don't think so. I finally discovered how my java code was doing it (rounding at the wrong place), and it was actually writing an odd number of bytes into the wav header.

**Comment by umlaeute** [comment]
well yes. both Pd and libsndfile agree that a soundfile shouldn't have a 'data' chunk that has an odd number of bytes.

but: where in the world does it say, that this is not allowed?
your test file is apparently broken, but a 24bit PCM file with 44001 samples should be valid and libsndfile still complains.

but yes, the two issues are different (Pd can read files with 44001 samples just fine), but related (both are assuming that the data-chunk must contain and odd number of bytes)
